### ISSUE TITLE
OpenSSL linking configuration mismatch between oqs (C) and oqs-sys

### ISSUE THREAD

**Issue by BlackHoleFox** [open-quantum-safe/liboqs-rust#17]
In the primary C library, linking to OpenSSL [is disabled on Windows](https://github.com/open-quantum-safe/liboqs/blob/fb5484cda68dac115dc563593c68388f89444b95/.CMake/alg_support.cmake#L13). However, in this crate, it is [enabled by default on all platforms](https://github.com/open-quantum-safe/liboqs-rust/blob/main/oqs-sys/build.rs#L44). Without `--no-default-features`, this causes a number of compilation errors on Windows when trying to build `oqs` in the build script.

Please consider not linking to OpenSSL by default, especially on Windows, although it causes installation headaches on all platforms.

**Comment by thomwiggers** [comment]
I need the equivalent of https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#platform-specific-dependencies for features but it does not seem to exist yet (https://github.com/rust-lang/cargo/issues/7914#issuecomment-776550578)

**Comment by BlackHoleFox** [comment]
> I need the equivalent of ...

To fix the OpenSSL issue on Windows, you can swap https://github.com/open-quantum-safe/liboqs-rust/blob/805ac8e8f36fa8926531ac4732dc99b29cc585e9/oqs-sys/build.rs#L45 out for 


```rust
if cfg!(all(not(windows), feature = "openssl")) {
```

Is there a reason why static linking is enabled by default? Is it needed for your use case of `liboqs-rust`? I am unable to get this crate to compile on Windows, even with the default features disabled (I have no idea how the Windows CI passes currently). I've also asked others to try and they've also encountered difficulties. Note that `liboqs` itself compiles and the binaries run fine in Visual Studio. I also tested the `liboqs.sln` project that `liboqs-sys` generates as part of the build process in Visual Studio and it compiles fine as well.

Every time I compile `liboqs-rust`, even with OpenSSL disabled, I'm met with a linker error:
```
Compiling oqs-sys v0.3.0
   Compiling oqs v0.3.0
error: could not find native static library `oqs`, perhaps an -L flag is missing?

error: aborting due to previous error

error: could not compile `oqs-sys`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

The only way the linking bug seems to fix itself is by changing this https://github.com/open-quantum-safe/liboqs-rust/blob/805ac8e8f36fa8926531ac4732dc99b29cc585e9/oqs-sys/build.rs#L56 to not statically link:
```rust
println!("cargo:rustc-link-lib=oqs");
```` 
Tl;dr I'm having a hard time seeing how something with the build script is not wrong currently. If the stuff above is acceptable and desired, I'll open a PR off [my testing branch](https://github.com/BlackHoleFox/liboqs-rust/tree/no-openssl-default).

**Comment by thomwiggers** [comment]
> if cfg!(all(not(windows), feature = "openssl")) {

I don't want to force-disable `openssl` on Windows, because if you _do_ set up everything correctly, it works fine, and gives you access to accelerated versions of a bunch of primitives that make up between 20% to up to 90% of the runtime of some schemes.

>I have no idea how the Windows CI passes currently

Presumably it's [set up differently](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md).

I'm not sure if `println!("cargo:rustc-link-lib=static=oqs");`  should indeed be static, I've mainly just copied that from some example somewhere initially and as I only run Linux, it's obviously not been a problem for me. I'd merge a PR that removes `static=` assuming it passes CI.

**Comment by thomwiggers** [comment]
> I don't want to force-disable openssl on Windows

I should add that if you don't want to have openssl in your subcrate, you can simply specify `no-default-features` and use a `build.rs` that emits the `openssl` feature where it is supported.

**Comment by BlackHoleFox** [comment]
> force disable OpenSSL on Windows

I don’t think it works at all from my testing. [This line](https://github.com/open-quantum-safe/liboqs/blob/main/src/common/rand/rand_nist.c#L35) in the primary C library explodes MSVC if you try to enable OpenSSL. And as stated originally, liboqs [disables compiling OpenSSL on Windows at all](https://github.com/open-quantum-safe/liboqs/blob/fb5484cda68dac115dc563593c68388f89444b95/.CMake/alg_support.cmake#L13). None of its CI jobs even attempt to enable it either.

> no-default-features

That would only be a solution if the static linking bug was fixed. But, due to the liboqs’ behavior with OpenSSL and Windows I think its not very important, at least until the C code is patched to support MSVC.

**Comment by thomwiggers** [comment]
>  This line in the primary C library explodes MSVC if you try to enable OpenSSL.

That file should not be compiled though...

**Comment by BlackHoleFox** [comment]
Well, it seems to be anyway. These are the errors that MSVC gives when you try to compile `main` on Windows with OpenSSL, sorry I didn't share these sooner:
```
 liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(36,15): error C2143: syntax error: missing ')' before '(' [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
  liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(36,24): error C2059: syntax error: ')' [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
  liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(36,25): error C2059: syntax error: ')' [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
  liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(37,13): error C2143: syntax error: missing ')' before 'type' [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
  liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(37,13): error C2091: function returns function [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
  liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(37,32): error C2085: 'handleErrors': not in formal parameter list [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
  liboqs-rust\oqs-sys\liboqs\src\common\rand\rand_nist.c(37,32): error C2143: syntax error: missing ';' before '{' [liboqs-rust\target\debug\build\oqs-sys-45a0b0771bd42382\out\build\src\common\common.vcxproj]
```
I tried changing the function to decorated like this and it looked to of fixed it.
```c
#ifdef OQS_USE_OPENSSL
# ifdef __GNUC__
__attribute__((noreturn))
# elif defined(_MSC_VER)
__declspec(noreturn)
```
  ```

**Comment by thomwiggers** [comment]
I'm running into problems if I don't statically link `liboqs` :(

**Comment by thomwiggers** [comment]
How are you building `liboqs`? You should be passing `-DOQS_BUILD_ONLY_LIB=Yes` if you want to match what this crate is doing.

**Comment by BlackHoleFox** [comment]
I’m not building it directly. I am attempting to compile it through this Rust crate. Every error I’ve posted has come the output of `oqs-sys`‘ build script using `cargo build`.

**Comment by thomwiggers** [comment]
That's definitely weird...

**Comment by thomwiggers** [comment]
I've figured out why CI was not having problems: CI wasn't running tests on Windows at all :see_no_evil: (#30)

**Comment by BlackHoleFox** [comment]
That would explain a bit. Can't believe I didn't see that sooner D:

**Comment by thomwiggers** [comment]
Yeah Github Action's workflow syntax is rather sensitive to these problems where you define matrix vars but don't actually use them

**Comment by thomwiggers** [comment]
The branch in #30 works on a Windows 10 Dev VM (from Microsoft) that's up-to-date; unfortunately I am having trouble updating the Github Actions VM to use a more recent Windows SDK.

**Comment by thomwiggers** [comment]
I've managed to figure out the problem in #30, so you could try that branch. Don't forget to make sure you update the submodule if you check out that branch.

**Comment by BlackHoleFox** [comment]
I checked out the branch and confirmed it fixes the linking issue. My tests also run after configuring the stack size.

**Comment by BlackHoleFox** [comment]
This is very unrelated, but it would be cool if we could get this `liboqs-rust` to work with the `openssl-sys` crate. To get this compiling, the `build.rs` logic has duplicated some of the searching and link logic of that crate. It also has a vendored feature to get around the headaches of needing OpenSSL present on your system.

Its probably not possible, at least not easily, since liboqs links to it, but it would be a great feature. I tried tinkering with it but the linker on both Linux and Windows isn't happy with the idea.

**Comment by thomwiggers** [comment]
The main problem is the `openssl-sys` crate `build.rs` does not expose the location where it found OpenSSL. Otherwise I could pass it on to `liboqs` — I did look into this.

**Comment by BlackHoleFox** [comment]
Ah, that sucks :/ Good to know though.
